lib/rouge/lexers/objective_c/common.rb

```diff
@@ -0,0 +1,184 @@
+# -*- coding: utf-8 -*- #
+# frozen_string_literal: true
+
+module Rouge
+  module Lexers
+    module ObjectiveCCommon
+      def self.at_keywords
+        @at_keywords ||= %w(
+          selector private protected public encode synchronized try
+          throw catch finally end property synthesize dynamic optional
+          interface implementation import
+        )
+      end
+
+      def self.at_builtins
+        @at_builtins ||= %w(true false YES NO)
+      end
+
+      def self.builtins
+        @builtins ||= %w(YES NO nil)
```

Comment:
If you have a look at the Objective-C visual sample on your branch versus master, you can see that `nil`, `YES` and `NO` are not being highlighted as `Name.Builtin` tokens.

I haven't thought very hard about this but I suspect that the reason is that the way that this works currently is that the method inherited from the C lexer calls this `self.builtins` method in the context of the Objective-C lexer. When you extract the common logic to the module, this context no longer seems to get picked up properly. It might be logic that needs to be duplicated in both instances (although that feels kludgey and like there should be a better way).

Reply:
Yes, I think the problem is that this line looks up `builtins` in the context of the class it's in:

https://github.com/rouge-ruby/rouge/blob/3ed2f30a41207d7c8fa43eb31ca91f02657fa9d1/lib/rouge/lexers/c.rb#L121

I've changed it to use reflection to add it to the class directly instead of passing it through the module, which I think solves the issue.

Are there any other discrepancies in the tests? I haven't really figured out a good way to compare them other than looking at them visually, and as you can see that isn't necessarily the best way to do this…

Reply:
Hmmm, I thought the block passed to `StateDSL#rule` was evaluated within the context of the defining class (in this case, `ObjectiveCpp`).

I wonder if the problem isn't the fact we're using Ruby's `Module#include` rather than `Object#extend`. Since we want to add the methods to the class mixing in the module it feels like `extend` might help (I'm not sure, though; am spitballing here). The `self` would need to be dropped from `ObjectiveCCommon` as `extend` will add the methods to the class object itself.

Reply:
Oh, if possible, please avoid force pushing changes when updating your branch. I'll squash all of the commits before merging so the final code change to Rouge's master branch will be clean. It makes it much easier for review if I can use GitHub's tools to see what's changed since a previous commit. Force pushing means those won't work.

Reply:
> Oh, if possible, please avoid force pushing changes when updating your branch.

Sure, if you're going to squash them on merge I'll make further changes in their own commits.

> I wonder if the problem isn't the fact we're using Ruby's `Module#include` rather than `Object#extend`. Since we want to add the methods to the class mixing in the module it feels like extend might help (I'm not sure, though; am spitballing here). The `self` would need to be dropped from `ObjectiveCCommon` as `extend` will add the methods to the class object itself.

OK, I converted the lexers to `extend` rather than `include` `ObjectiveCCommon` and moved the methods back to how they were before. I think they're still being picked up correctly?

Reply:
Woo! It looks like that worked. When I run `rackup` to test the visual samples, I see the `nil`, `YES` and `NO` keywords being tokenised properly :) There appear to be no other visual differences between the samples on `master` and this PR. I think that might have solved the problem 👍